Add react tip button

shapkarin opened the pull request:
closes #21

davidyuk commented on src/button-react/index.js:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
Mostly probably that we won't add any files to button-react, so I propose to rename this file to `src/button-react.js`.

shapkarin replied:
thanks, next it can be replaced with folder is that needed

davidyuk commented on package.json:
```diff
@@ -4,7 +4,10 @@
   "main": "dist/index.js",
   "browser": "dist/index.js",
-  "dependencies": {},
+  "dependencies": {
```
So, what about the difference between `peerDependencies` and `optionalDependencies`? Which of them suits better in our case?
https://github.com/aeternity/superhero-button-react/issues/2#issuecomment-656546171

shapkarin replied:
optionalDependencies – when package was not installed in some reason other packages will be installed and developer will not see the error.
peerDependencies – when package is required.

shapkarin replied:
@davidyuk I think to leave it as it is

davidyuk replied:
`peerDependencies` suits better in our case than usual `dependencies` because npm won't allow installing two different versions of `react` (for example) as it may do using usual `dependencies`:

> Trying to install another plugin with a conflicting requirement will cause an error. 

https://docs.npmjs.com/cli/v6/configuring-npm/package-json#peerdependencies

`optionalDependencies` suits better in our case than `peerDependencies` because we don't need npm to warn us about missing `react` package (as it does on `peerDependencies`) if we are using `superhero-utils` in a project that uses a different UI framework.

davidyuk commented on src/button-react/index.js:
```diff
@@ -0,0 +1,22 @@
+  size: PropTypes.oneOf(['icon', 'small', 'medium', 'large']),
+  url: PropTypes.string,
+  account: PropTypes.string,
+};
```
Please, setup your IDE to handle [.editorconfig](https://github.com/aeternity/superhero-utils/blob/b58a16c41267c97d5d7ee9c0beee4f692290c891/.editorconfig#L9) files.

shapkarin replied:
done, thanks

davidyuk reviewed (commented):
I don't think that all environments will support es6 imports, classes, and arrow functions, so proposing to pass button-react through webpack.

Commit:
add react tip button

davidyuk commented:
@shapkarin Any response to the first comment about es6 features and webpack?

shapkarin commented:
> @shapkarin Any response to the first comment about es6 features and webpack?

@davidyuk I will add

davidyuk commented on button-react.js:
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
I think this file should be in `src` folder

shapkarin replied:
right

davidyuk commented on webpack.config.js:
```diff
@@ -40,4 +40,5 @@ const genConfig = (filename, { inlineCss } = {}) => ({
 module.exports = [
   genConfig('index-without-styles.js'),
   genConfig('index.js', { inlineCss: true }),
+  genConfig('button-react.js'),
```
This is an output file name, have you specified the input one? The built `button-react.js` is equal to `index-without-styles.js`, it doesn't contain a React version of Superhero button.

shapkarin replied:
Got that

shapkarin replied:
Done

davidyuk reviewed (changes requested):
Document this feature in the README, please.

shapkarin commented:
> Document this feature in the README, please.

Added basic version.

davidyuk commented on README.md:
```diff
@@ -22,12 +22,16 @@ all, and write your own instead.
 ## Usage
 
+- [Plain JavaScript](#Button-(`superheroUtils.createButton`))
+- [React.js](#React.js-component)
```
These links are not working on GitHub

davidyuk replied:
@shapkarin can you fix this?

shapkarin replied:
@davidyuk i fixed already and pushed, they are fine.

davidyuk commented on README.md:
```diff
@@ -22,12 +22,16 @@ all, and write your own instead.
   (this option simplifies integration into Frontend frameworks like Vue/React)
 - options object
 
+#  
```
What is this?

davidyuk commented on README.md:
````diff
@@ -79,6 +83,19 @@ Additional examples can be found [here](./index.html).
+```js
+import TipButton from '@aeternity/superhero-utils/button-react';
+
+<TipButton size="icon" account="AK_addressThatYouWonder">
````
```suggestion
<TipButton size="icon" account="example.chain">
```
Proposing to use the consistent data between examples, before was used `example.chain`

davidyuk replied:
Also, let's name it just Button as it named above

davidyuk commented on package.json:
```diff
@@ -4,7 +4,6 @@
   "description": "",
   "main": "dist/index.js",
   "browser": "dist/index.js",
-  "dependencies": {},
```
Try to avoid useless changes

shapkarin replied:
It's not useless, please have a look at master branch https://github.com/aeternity/superhero-utils/blob/2ab1d46ff3dd03457e89a67fba3d6374fd729984/package.json#L7

davidyuk commented on webpack.config.js:
```diff
@@ -2,13 +2,16 @@ const path = require('path');
     library: 'superheroUtils',
     libraryTarget: 'umd'
   },
+  entry: {
+    [inputName.slice(0, -3)]: `./src/${inputName}`
+  },
```
```suggestion
  entry: `./src/${inputName}`,
```
https://webpack.js.org/concepts/entry-points/#single-entry-shorthand-syntax

davidyuk replied:
Why don't do exactly as I proposed?

shapkarin replied:
Not sure what do you mean

I changed it to the 
```js
    entry: `./src/${inputName}`,
```
as I understand it's exactly as you proposed

davidyuk commented on webpack.config.js:
```diff
@@ -2,13 +2,16 @@ const path = require('path');
   output: {
     path: path.resolve(__dirname, 'dist'),
-    filename,
+    filename: outputName,
```
```suggestion
    filename: outputFilename,
```
like, to don't make names shorter than webpack does🤷‍♀️

shapkarin replied:
Ok, why not, more semantic

davidyuk commented on package.json:
```diff
@@ -27,6 +26,10 @@
     "webpack-cli": "^3.3.12"
   },
+  "optionalDependencies": {
+    "prop-types": "^15.7.2",
+    "react": "^17.0.0"
```
Currently, these dependencies are inlined into the bundle, let's use [externals](https://webpack.js.org/configuration/externals/) to keep them extracted.

shapkarin replied:
I hope that will help with that, I will add it
```js
  externals: {
    'prop-types': 'prop-types',
    'react': 'react'
  },
```

davidyuk replied:
The general idea is not to hope but to test it using a bundle analyzer and trying to use it in different environments

shapkarin replied:
I will check it at Monday's morning

shapkarin commented:
@davidyuk squashed into one commit, please check this PR

davidyuk commented on webpack.config.js:
```diff
@@ -2,14 +2,21 @@ const path = require('path');
 const { CleanWebpackPlugin } = require('clean-webpack-plugin');
 
-const genConfig = (filename, { inlineCss } = {}) => ({
+const genConfig = (outputFilename, { inlineCss, inputName = 'index.js' } = {}) => ({
```
```suggestion
const genConfig = (outputFilename, { inlineCss, inputFilename = 'index.js' } = {}) => ({
```
🤷‍♀️

davidyuk reviewed (commented):
Not all issues from the last review are resolved

Commit:
fix README.md anchor links and genConfig option name

Update webpack.config.js

Co-authored-by: Denis Davidyuk <denis_davidyuk@hotmail.com>

again temporary commit just to track

fix README.md anchor links

fix genConfig option name

davidyuk commented on webpack.config.js:
```diff
@@ -2,14 +2,21 @@ const path = require('path');
     library: 'superheroUtils',
     libraryTarget: 'umd'
   },
+  entry: {
+    entry: `./src/${inputFilename}`,
+  },
```
Let's try one more time:
```suggestion
  entry: `./src/${inputFilename}`,
```

shapkarin replied:
thanks

davidyuk commented on webpack.config.js:
```diff
@@ -40,4 +47,5 @@ const genConfig = (filename, { inlineCss } = {}) => ({
 module.exports = [
   genConfig('index-without-styles.js'),
   genConfig('index.js', { inlineCss: true }),
+  genConfig('button-react.js', { inputName: 'button-react.js'}),
```
```suggestion
  genConfig('button-react.js', { inputFilename: 'button-react.js'}),
```

davidyuk reviewed (commented):
In general, this PR looks fine, but there is one thing related to the documentation that I would like to check before merging this. I will come back later to this PR.

Commit:
Update webpack.config.js

Co-authored-by: Denis Davidyuk <denis_davidyuk@hotmail.com>

shapkarin commented:
@davidyuk fixed and I clean build folder and run it once again, looks that all is fine

davidyuk commented on src/button-react.js:
```diff
@@ -0,0 +1,22 @@
+    super(props);
+    this.button = React.createRef();
+    this.componentDidMount = this.componentDidUpdate = () =>
+      createButton(this.button.current, this.props);
```
While working on #38 I have found that the replacement of the DOM node breaks subsequent component updates. I have fixed it there by using a different API.

davidyuk commented on webpack.config.js:
```diff
@@ -2,14 +2,19 @@ const path = require('path');
   target: 'web',
+  externals: {
+    'prop-types': 'prop-types',
+    'react': 'react'
```
Actually, this is not working if imported using the `script` tag because React exposes itself there as `React`, it is fixed in #38

davidyuk reviewed (commented):
In #38 I have implemented a more general approach with additional examples allowing to test that. Proposing to close this and merge that instead.
